### PR TITLE
feat(neptune): high level constructs for db clusters and instances

### PR DESCRIPTION
**Pull request by cornerwings**
- This change adds higher level constructs for Neptune clusters
- Adds higher-level constructs for
  - AWS::Neptune::DBCluster
  - AWS::Neptune::DBInstance
  - AWS::Neptune::DBClusterParameterGroup
  - AWS::Neptune::DBParameterGroup
  - AWS::Neptune::DBSubnetGroup

fixes aws#12762

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by njlynch — COMMENTED** [review]
Wow, thanks for this contribution!

It's going to take me a few days to have the time to dive deep into this, but I've given it a quick skim and it looks like it's in really great shape. I imagine this is heavily inspired/borrowed from the RDS module.

One point of note is that the description claims this provides the high-level constructs for clusters and instances, but I'm only seeing "L2s" for the cluster. That's totally fine; this is big enough as-is, and we should be able to add in the instances later. I just wanted to clarify intent.

Another minor high-level note is that the pattern of `cluster-ref` and `cluster` is kind of a "legacy" holdover and I'd rather us just combine into one file. Same with `props`, although I can imagine the instance-specific stuff there going into the (eventual) instance L2.

Lastly, we'll want to add more detail to the README to cover some of the other usage/elements. It doesn't need to be 100% comprehensive, but a bit more than is currently there I think is necessary to get folks up and running quickly.

I'll circle back around (hopefully) next week to do the first pass on the code itself.

Thanks again!

**Comment by cornerwings** [comment]
Thanks Nick for taking a quick pass at this!

Good point regarding the instances - I was initially thinking of adding a L2 construct for that but didn't end up doing it because most customers just pick the no. of instances and let the service figure out the distribution. But it might be worth adding if someone wants to give specific names to instances or launch them in specific availability zones. I will take a stab at this, will get rid of current `InstanceProps` and merge them back to `ClusterProps` to avoid confusion with the actual instance props (which will take some additional attributes such as reference to the cluster). Thanks for the note on `cluster-ref` pattern, will clean that up too along with adding more examples in README.

File: packages/@aws-cdk/aws-neptune/test/endpoint.test.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import { Endpoint } from '../lib';
+
+// A numeric CDK token (see: https://docs.aws.amazon.com/cdk/latest/guide/tokens.html#tokens_number)
+const CDK_NUMERIC_TOKEN = -1.8881545897087626e+289;
```
Comment (njlynch) [inline packages/@aws-cdk/aws-neptune/test/endpoint.test.ts:4]: Agh! I see you've copy/pasted this, but let's not repeat this. We can use the `Token` class to deal with tokens. This test otherwise isn't actually testing what you think it does.

File: packages/@aws-cdk/aws-neptune/test/endpoint.test.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+describe('Endpoint', () => {
+  test('accepts tokens for the port value', () => {
```
Comment (njlynch) [inline packages/@aws-cdk/aws-neptune/test/endpoint.test.ts:7]: _If_ we want to keep this test, I think the right approach would be to use `Token.asAny`, and verifying that the port is an unresolved token.

**Review by njlynch — DISMISSED** [review]
Looks good! I've done a first pass of comments throughout.

File: packages/@aws-cdk/aws-neptune/lib/cluster.ts [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+    ];
+
+    // Default to encrypted storage
+    const storageEncrypted = props.storageEncrypted ?? true;
+
+    if (props.kmsKey && !storageEncrypted) {
+      throw new Error('KMS key supplied but storageEncrypted is false');
+    }
+
+    // Deletion protection
```
[the remark this answers is not in the document]
Comment (cornerwings) [reply packages/@aws-cdk/aws-neptune/lib/cluster.ts:391]: Removed the comment

File: packages/@aws-cdk/aws-neptune/lib/cluster.ts [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+   * Reader endpoint address
+   */
+  readonly readerEndpointAddress: string;
+}
+
+
+/**
+ * A new or imported clustered database.
+ */
+abstract class DatabaseClusterBase extends Resource implements IDatabaseCluster {
```
[the remark this answers is not in the document]
Comment (cornerwings) [reply packages/@aws-cdk/aws-neptune/lib/cluster.ts:244]: ```DatabaseClusterBase``` gets used when importing an existing cluster, see ```DatabaseCluster.fromDatabaseClusterAttributes``` method. Any suggestions on how to refactor this if we don't have base classes? Originally I was also planning to add another class for clusters that are restored from snapshot but wanted to first finish the basic setup before taking that on.

File: packages/@aws-cdk/aws-neptune/lib/instance.ts [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+}
+
+/**
+ * A new or imported database instance.
+ */
+abstract class DatabaseInstanceBase extends cdk.Resource implements IDatabaseInstance {
+  /**
+   * Import an existing database instance.
+   */
+  public static fromDatabaseInstanceAttributes(scope: Construct, id: string, attrs: DatabaseInstanceAttributes): IDatabaseInstance {
```
[the remark this answers is not in the document]
Comment (cornerwings) [reply packages/@aws-cdk/aws-neptune/lib/instance.ts:124]: Moved this to `DatabaseInstance` and removed the base class

File: packages/@aws-cdk/aws-neptune/lib/cluster.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (cornerwings) [reply packages/@aws-cdk/aws-neptune/lib/cluster.ts:244]: Removed the base class, originally added this to also support database clusters that are restored from snapshot but we can add this when it is needed

File: packages/@aws-cdk/aws-neptune/lib/instance.ts [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+ */
+export class DatabaseInstance extends DatabaseInstanceBase implements IDatabaseInstance {
+  /**
+   * The instance's database cluster
+   */
+  public readonly cluster: IDatabaseCluster;
+
+  /**
+   * @inheritdoc
+   */
```
[the remark this answers is not in the document]
Comment (cornerwings) [reply packages/@aws-cdk/aws-neptune/lib/instance.ts:212]: Removed this class

File: packages/@aws-cdk/aws-neptune/package.json [hunk 7]
```diff
@@ -74,22 +74,36 @@
...
     "constructs": "^3.2.0"
   },
   "engines": {
     "node": ">= 10.13.0 <13 || >=13.7.0"
   },
+  "awslint": {
+    "exclude": [
+      "props-physical-name:@aws-cdk/aws-neptune.ParameterGroupProps",
+      "props-physical-name:@aws-cdk/aws-neptune.ClusterParameterGroupProps",
+      "props-physical-name:@aws-cdk/aws-neptune.SubnetGroupProps"
```
[the remark this answers is not in the document]
Comment (cornerwings) [reply packages/@aws-cdk/aws-neptune/package.json:102]: Yeah its strange, if I remove these - this is the error I get during build.

```
error: [awslint:props-physical-name:@aws-cdk/aws-neptune.ClusterParameterGroupProps] Every Resource must have a single physical name construction property, with a name that is an ending substring of <cfnResource>Name
error: [awslint:props-physical-name:@aws-cdk/aws-neptune.ParameterGroupProps] Every Resource must have a single physical name construction property, with a name that is an ending substring of <cfnResource>Name
error: [awslint:props-physical-name:@aws-cdk/aws-neptune.SubnetGroupProps] Every Resource must have a single physical name construction property, with a name that is an ending substring of <cfnResource>Name
```

File: packages/@aws-cdk/aws-neptune/test/endpoint.test.ts (hunk 1 above)
Comment (cornerwings) [reply packages/@aws-cdk/aws-neptune/test/endpoint.test.ts:4]: Removed this test altogether

File: packages/@aws-cdk/aws-neptune/lib/cluster.ts [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+   */
+  public readonly instanceEndpoints: Endpoint[] = [];
+
+  constructor(scope: Construct, id: string, props: DatabaseClusterProps) {
+    super(scope, id);
+
+    this.vpc = props.vpc;
+    this.vpcSubnets = props.vpcSubnets ? props.vpcSubnets : {
+      subnetType: ec2.SubnetType.PRIVATE,
+    };
```
[the remark this answers is not in the document]
Comment (cornerwings) [reply packages/@aws-cdk/aws-neptune/lib/cluster.ts:360]: Done

File: packages/@aws-cdk/aws-neptune/lib/cluster.ts [hunk 9]
```diff
@@ -0,0 +1,463 @@
...
+
+  /**
+   * The subnets used by the DB subnet group.
+   */
+  public readonly vpcSubnets: ec2.SubnetSelection;
+
+  /**
+   * The security groups used by DB cluster.
+   */
+  protected readonly securityGroups: ec2.ISecurityGroup[];
```
[the remark this answers is not in the document]
Comment (cornerwings) [reply packages/@aws-cdk/aws-neptune/lib/cluster.ts:337]: Removed this member from the cluster class

File: packages/@aws-cdk/aws-neptune/lib/cluster.ts [hunk 10]
```diff
@@ -0,0 +1,463 @@
...
+   * Whether to enable storage encryption
+   *
+   * @default true
+   */
+  readonly storageEncrypted?: boolean;
+
+  /**
+   * Number of Neptune compute instances
+   *
+   * @default 1
```
[the remark this answers is not in the document]
Comment (cornerwings) [reply packages/@aws-cdk/aws-neptune/lib/cluster.ts:83]: Don't have a particular preference, picked one here to allow for cases where someone wanted customization over names of the instances (use cluster to name the writer and then use instance to name each of the reader), but this is purely syntactic as the names don't mean anything ultimately.

Please let me know if you think it is better to default to two instead.

File: packages/@aws-cdk/aws-neptune/lib/cluster.ts [hunk 11]
```diff
@@ -0,0 +1,463 @@
...
+/**
+ * Properties for a new database cluster
+ */
+export interface DatabaseClusterProps {
+  /**
+   * What version of the database to start
+   *
+   * @default - The default engine version.
+   */
+  readonly engineVersion?: string;
```
[the remark this answers is not in the document]
Comment (cornerwings) [reply packages/@aws-cdk/aws-neptune/lib/cluster.ts:48]: I thought about this one, engine version releases are done pretty frequently and that might introduce additional friction if this is modeled as enum. Here is the list of engine releases for Neptune: https://docs.aws.amazon.com/neptune/latest/userguide/engine-releases.html#engine-releases-versions

Any thoughts on leaving this as a string field?

File: packages/@aws-cdk/aws-neptune/lib/cluster.ts [hunk 12]
```diff
@@ -0,0 +1,463 @@
...
+import { ISubnetGroup, SubnetGroup } from './subnet-group';
+
+/**
+ * Backup configuration for Neptune databases
+ *
+ * @default - The retention period for automated backups is 1 day.
+ * The preferred backup window will be a 30-minute window selected at random
+ * from an 8-hour block of time for each AWS Region.
+ */
+export interface BackupProps {
```
[the remark this answers is not in the document]
Comment (cornerwings) [reply packages/@aws-cdk/aws-neptune/lib/cluster.ts:19]: Thanks, merged these into the `DatabaseClusterProps`

File: packages/@aws-cdk/aws-neptune/README.md [hunk 13]
```diff
@@ -9,10 +9,93 @@
...
+> They are subject to non-backward compatible changes or removal in any future version. These are
+> not subject to the [Semantic Versioning](https://semver.org/) model and breaking changes will be
+> announced in the release notes. This means that while you may use them, you may need to update
+> your source code when upgrading to a newer version of this package.
+
 ---
 
 <!--END STABILITY BANNER-->
 
+## Starting a Neptune Database
```
[the remark this answers is not in the document]
Comment (cornerwings) [reply packages/@aws-cdk/aws-neptune/README.md:24]: Makes sense, stole the intro from Amazon Neptune documentation and placed it here.

File: packages/@aws-cdk/aws-neptune/README.md [hunk 14]
```diff
@@ -9,10 +9,93 @@
...
+
+To set up a Neptune database, define a `DatabaseCluster`. You must always launch a database in a VPC. 
+
+```ts
+const cluster = new DatabaseCluster(this, 'Database', {
```
[the remark this answers is not in the document]
Comment (cornerwings) [reply packages/@aws-cdk/aws-neptune/README.md:29]: This is great, added fixtures and ran `yarn rosetta:extract --strict` to verify everything works as expected.

File: packages/@aws-cdk/aws-neptune/package.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (njlynch) [reply packages/@aws-cdk/aws-neptune/package.json:102]: Ah. The test is checking against the Cfn resource name, so it's expecting `dbSubnetGroupName` (for example) rather than `subnetGroupName`. In that case, these exceptions are fine.

File: packages/@aws-cdk/aws-neptune/lib/cluster.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (njlynch) [reply packages/@aws-cdk/aws-neptune/lib/cluster.ts:66]: We can start with 1, can always change it based on community feedback.

File: packages/@aws-cdk/aws-neptune/lib/cluster.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (njlynch) [reply packages/@aws-cdk/aws-neptune/lib/cluster.ts:21]: The enum-like pattern (linked above) does provide some options there, as we model known values, but don't block new/custom values from customers. Ideally, we'd auto-generate this code and update the values as new engines are released, but we haven't (yet) built up a pattern/infrastructure for that.

The main point being that modelling it allows customers to quickly discover (at least a slightly out-of-date subset of the) valid values for the field, whereas it being a string effectively leaves customers guessing, especially as the CloudFormation docs are a bit unhelpful.

For example, which of following is expected/valid: `1.0.4.0`, `1.0.4.0.R2`, `Neptune 1.0.4.0.R2`? I highly _suspect_ it's the first one, but I can't be 100% sure until I try.

File: packages/@aws-cdk/aws-neptune/lib/cluster.ts [hunk 15]
```diff
@@ -0,0 +1,412 @@
...
+   * The port the Neptune cluster will listen on
+   *
+   * @default - The default engine port
+   */
+  readonly port?: number;
+
+  /**
+   * How many days to retain the backup
+   *
+   * @default - Backup retention period for automated backups is 1 day
```
Comment (njlynch) [inline packages/@aws-cdk/aws-neptune/lib/cluster.ts:33]:
```suggestion
   * @default cdk.Duration.days(1)
```

File: packages/@aws-cdk/aws-neptune/lib/cluster.ts [hunk 16]
```diff
@@ -113,13 +96,6 @@ export interface DatabaseClusterProps {
    */
   readonly associatedRoles?: iam.IRole[];
 
-  /**
-   * Indicates whether the DB cluster is IAM auth enabled.
-   *
-   * @default - false, connections to database don't require IAM auth
-   */
-  readonly iamAuthEnabled?: boolean;
```
Comment (njlynch) [inline packages/@aws-cdk/aws-neptune/lib/cluster.ts]: Why was this removed?

File: packages/@aws-cdk/aws-neptune/lib/cluster.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (cornerwings) [reply packages/@aws-cdk/aws-neptune/lib/cluster.ts:21]: Ah makes sense, created a new enum-like class for engine version and added corresponding constants in there.

File: packages/@aws-cdk/aws-neptune/README.md [hunk 17]
```diff
@@ -9,10 +9,100 @@
...
 <!--END STABILITY BANNER-->
 
-```ts
+Amazon Neptune is a fast, reliable, fully managed graph database service that makes it easy to build and run applications that work with highly connected datasets. The core of Neptune is a purpose-built, high-performance graph database engine. This engine is optimized for storing billions of relationships and querying the graph with milliseconds latency. Neptune supports the popular graph query languages Apache TinkerPop Gremlin and W3C’s SPARQL, enabling you to build queries that efficiently navigate highly connected datasets.
+
+The `@aws-cdk/aws-neptune` package contains primitives for setting up Neptune database clusters and instances.
+
+```ts nofixture
 import * as neptune from '@aws-cdk/aws-neptune';
+
```
[the remark this answers is not in the document]
Comment (cornerwings) [reply packages/@aws-cdk/aws-neptune/README.md:30]: Fixed

File: packages/@aws-cdk/aws-neptune/lib/cluster.ts (hunk 15 above)
Comment (cornerwings) [reply packages/@aws-cdk/aws-neptune/lib/cluster.ts:33]: Done

File: packages/@aws-cdk/aws-neptune/lib/cluster.ts (hunk 16 above)
Comment (cornerwings) [reply packages/@aws-cdk/aws-neptune/lib/cluster.ts]: I wanted to test this before adding support for it, found some conflicting information where AWS CLI says IAM authentication is not supported by Neptune at the cluster level but is supported at instance level (whereas CFN has no flag to pass this value at instance level). Wanted to check if SDK documentation is incorrect or if it is a really an instance specific setting and if so whether CFN support for IAM needs to be fixed.

**Review by njlynch — APPROVED** [review]
Looks good. Looking forward to getting these out there and getting some feedback!

Thanks for the contribution!
